File: tools/art-engine/src/core/primitives.ts
/**
 * Pixel-level shape drawing primitives.
 *
 * All algorithms work at integer coordinates — no anti-aliasing.
 * Bresenham's line, midpoint ellipse, scanline fill.
 */

import { type PixelGrid, type RGBA, rgbaEqual } from './pixel-grid.js';

/** Bresenham's line algorithm. Draws from (x0,y0) to (x1,y1) inclusive. */
export function drawLine(
  grid: PixelGrid,
  x0: number, y0: number,
  x1: number, y1: number,
  color: RGBA,
): void {
  let dx = Math.abs(x1 - x0);
  let dy = -Math.abs(y1 - y0);
  const sx = x0 < x1 ? 1 : -1;
  const sy = y0 < y1 ? 1 : -1;
  let err = dx + dy;

  let cx = x0;
  let cy = y0;

  for (;;) {
    grid.setPixel(cx, cy, color);
    if (cx === x1 && cy === y1) break;
    const e2 = 2 * err;
    if (e2 >= dy) {
      err += dy;
      cx += sx;
    }
    if (e2 <= dx) {
      err += dx;
      cy += sy;
    }
  }
}

/**
 * Draw a rectangle.
 * (x, y) is top-left corner. w and h are dimensions.
 * If filled, fills the entire w x h area. If not, draws only the border.
 */
export function drawRect(
  grid: PixelGrid,
  x: number, y: number,
  w: number, h: number,
  color: RGBA,
  filled = false,
): void {
  if (filled) {
    for (let py = y; py < y + h; py++) {
      for (let px = x; px < x + w; px++) {
        grid.setPixel(px, py, color);
      }
    }
  } else {
    // Top and bottom edges
    for (let px = x; px < x + w; px++) {
      grid.setPixel(px, y, color);
      grid.setPixel(px, y + h - 1, color);
    }
    // Left and right edges
    for (let py = y; py < y + h; py++) {
      grid.setPixel(x, py, color);
      grid.setPixel(x + w - 1, py, color);
    }
  }
}

/**
 * Midpoint ellipse algorithm.
 * (cx, cy) is center, rx/ry are radii.
 */
export function drawEllipse(
  grid: PixelGrid,
  cx: number, cy: number,
  rx: number, ry: number,
  color: RGBA,
  filled = false,
): void {
  if (rx <= 0 || ry <= 0) return;

  let x = 0;
  let y = ry;
  const rx2 = rx * rx;
  const ry2 = ry * ry;
  const twoRx2 = 2 * rx2;
  const twoRy2 = 2 * ry2;
  let px = 0;
  let py = twoRx2 * y;

  // Region 1
  let p = Math.round(ry2 - rx2 * ry + 0.25 * rx2);
  while (px < py) {
    if (filled) {
      drawHLine(grid, cx - x, cx + x, cy + y, color);
      drawHLine(grid, cx - x, cx + x, cy - y, color);
    } else {
      plotEllipsePoints(grid, cx, cy, x, y, color);
    }
    x++;
    px += twoRy2;
    if (p < 0) {
      p += ry2 + px;
    } else {
      y--;
      py -= twoRx2;
      p += ry2 + px - py;
    }
  }

  // Region 2
  p = Math.round(ry2 * (x + 0.5) * (x + 0.5) + rx2 * (y - 1) * (y - 1) - rx2 * ry2);
  while (y >= 0) {
    if (filled) {
      drawHLine(grid, cx - x, cx + x, cy + y, color);
      drawHLine(grid, cx - x, cx + x, cy - y, color);
    } else {
      plotEllipsePoints(grid, cx, cy, x, y, color);
    }
    y--;
    py -= twoRx2;
    if (p > 0) {
      p += rx2 - py;
    } else {
      x++;
      px += twoRy2;
      p += rx2 - py + px;
    }
  }
}

function plotEllipsePoints(
  grid: PixelGrid, cx: number, cy: number,
  x: number, y: number, color: RGBA,
): void {
  grid.setPixel(cx + x, cy + y, color);
  grid.setPixel(cx - x, cy + y, color);
  grid.setPixel(cx + x, cy - y, color);
  grid.setPixel(cx - x, cy - y, color);
}

function drawHLine(grid: PixelGrid, x0: number, x1: number, y: number, color: RGBA): void {
  for (let x = x0; x <= x1; x++) {
    grid.setPixel(x, y, color);
  }
}

/**
 * Flood fill using a queue (BFS). Fills connected region of target color with fill color.
 */
export function floodFill(grid: PixelGrid, startX: number, startY: number, fillColor: RGBA): void {
  if (!grid.isInBounds(startX, startY)) return;

  const targetColor = grid.getPixel(startX, startY);
  if (rgbaEqual(targetColor, fillColor)) return;

  const queue: [number, number][] = [[startX, startY]];
  const visited = new Set<number>();
  const key = (x: number, y: number): number => y * grid.width + x;

  visited.add(key(startX, startY));

  while (queue.length > 0) {
    const [x, y] = queue.shift()!;
    grid.setPixel(x, y, fillColor);

    for (const [nx, ny] of [[x - 1, y], [x + 1, y], [x, y - 1], [x, y + 1]] as const) {
      if (!grid.isInBounds(nx, ny)) continue;
      const k = key(nx, ny);
      if (visited.has(k)) continue;
      if (!rgbaEqual(grid.getPixel(nx, ny), targetColor)) continue;
      visited.add(k);
      queue.push([nx, ny]);
    }
  }
}

File: tools/art-engine-v2/src/core/primitives.py
"""Pixel-art drawing primitives: lines, rectangles, ellipses, flood fill."""
from __future__ import annotations
from collections import deque
import numpy as np

Color = tuple[int, int, int, int]  # RGBA


def _set_pixel(canvas: np.ndarray, x: int, y: int, color: Color) -> None:
    """Set pixel with bounds checking."""
    h, w = canvas.shape[:2]
    if 0 <= x < w and 0 <= y < h:
        canvas[y, x] = color


def draw_line(canvas: np.ndarray, x0: int, y0: int, x1: int, y1: int, color: Color) -> None:
    """Bresenham's line algorithm."""
    # implement full Bresenham with octant handling
    dx = abs(x1 - x0)
    dy = -abs(y1 - y0)
    sx = 1 if x0 < x1 else -1
    sy = 1 if y0 < y1 else -1
    err = dx + dy
    while True:
        _set_pixel(canvas, x0, y0, color)
        if x0 == x1 and y0 == y1:
            break
        e2 = 2 * err
        if e2 >= dy:
            err += dy
            x0 += sx
        if e2 <= dx:
            err += dx
            y0 += sy


def draw_rect(canvas: np.ndarray, x0: int, y0: int, x1: int, y1: int, color: Color) -> None:
    """Draw outline rectangle."""
    draw_line(canvas, x0, y0, x1, y0, color)  # Top
    draw_line(canvas, x0, y1, x1, y1, color)  # Bottom
    draw_line(canvas, x0, y0, x0, y1, color)  # Left
    draw_line(canvas, x1, y0, x1, y1, color)  # Right


def draw_filled_rect(canvas: np.ndarray, x0: int, y0: int, x1: int, y1: int, color: Color) -> None:
    """Draw filled rectangle."""
    h, w = canvas.shape[:2]
    for y in range(max(0, min(y0, y1)), min(h, max(y0, y1) + 1)):
        for x in range(max(0, min(x0, x1)), min(w, max(x0, x1) + 1)):
            canvas[y, x] = color


def draw_ellipse(canvas: np.ndarray, cx: int, cy: int, rx: int, ry: int, color: Color) -> None:
    """Midpoint ellipse algorithm."""
    x = 0
    y = ry
    rx2 = rx * rx
    ry2 = ry * ry

    # Region 1
    p1 = ry2 - rx2 * ry + 0.25 * rx2
    dx = 2 * ry2 * x
    dy = 2 * rx2 * y

    while dx < dy:
        _set_pixel(canvas, cx + x, cy + y, color)
        _set_pixel(canvas, cx - x, cy + y, color)
        _set_pixel(canvas, cx + x, cy - y, color)
        _set_pixel(canvas, cx - x, cy - y, color)
        x += 1
        dx += 2 * ry2
        if p1 < 0:
            p1 += dx + ry2
        else:
            y -= 1
            dy -= 2 * rx2
            p1 += dx - dy + ry2

    # Region 2
    p2 = ry2 * (x + 0.5) ** 2 + rx2 * (y - 1) ** 2 - rx2 * ry2
    while y >= 0:
        _set_pixel(canvas, cx + x, cy + y, color)
        _set_pixel(canvas, cx - x, cy + y, color)
        _set_pixel(canvas, cx + x, cy - y, color)
        _set_pixel(canvas, cx - x, cy - y, color)
        y -= 1
        dy -= 2 * rx2
        if p2 > 0:
            p2 += rx2 - dy
        else:
            x += 1
            dx += 2 * ry2
            p2 += dx - dy + rx2


def flood_fill(canvas: np.ndarray, x: int, y: int, color: Color) -> None:
    """BFS flood fill from (x, y) with given color."""
    h, w = canvas.shape[:2]
    if not (0 <= x < w and 0 <= y < h):
        return
    target = tuple(canvas[y, x])
    fill = tuple(color)
    if target == fill:
        return
    queue = deque([(x, y)])
    visited = set()
    while queue:
        cx, cy = queue.popleft()
        if (cx, cy) in visited:
            continue
        if not (0 <= cx < w and 0 <= cy < h):
            continue
        if tuple(canvas[cy, cx]) != target:
            continue
        visited.add((cx, cy))
        canvas[cy, cx] = color
        queue.extend([(cx+1, cy), (cx-1, cy), (cx, cy+1), (cx, cy-1)])
